neomodel/util.py

```diff
@@ -75,6 +80,21 @@ def __str__(self):
         return "\n".join(ncr_items)
 
 
+class Path(NeoPath):
+    def __init__(self, nodes, *relationships):
+        assert all([isinstance(node, BaseNode) for node in nodes])
+        for i, relationship in enumerate(relationships, start=1):
+            assert isinstance(relationship, BaseRel)
```

Comment:
This check, along with the one on line 85 are possibly redundant. What else than nodes and relationships can the retrieved objects be here?

(Also, this kind of check will at some point be superseded by [type checking](https://github.com/neo4j-contrib/neomodel/issues/473)

Reply:
I drew a lot of inspiration from the implementation of the neo4j `Path` class: https://github.com/neo4j/neo4j-python-driver/blob/5.0/src/neo4j/graph/__init__.py#L297. This might not be the best idea, I don't know.
On the other hand, I couldn't test that they were `StructuredNode` and `StructuredRel` because that was causing a cyclic import error. That's the reason I made base classes for both types. 
I can remove them if you prefer.

Reply:
Hey @aanastasiou,

Have you seen my reply?

Cheers

Reply:
@aanastasiou This is the last standing point for this PR, any opinion on this ?

Reply:
I agree with removing this pseudo-type.

However I have no clue about the different db versions. I am completely new to neo4j and may need help identifying which versions should be tested. 

I haven't implemented any tests for the moment.

What should I do now?

Reply:
The tests are running once you push your PR. The whole battery of tests runs on each different version of neo4j.

Tests use `pytest`. Have a look at that. Also, the test needs a docker instance running the neo4j db of choice.

So, whether locally or when pushing the PR, the tests will run and we will be able to see the result.

In terms of what we are doing next:

1. Please go ahead and change the code as discussed
2. Is this more or less what you are trying to do?
    ```
          z = neomodel.db.cypher_query(" match p = (a:Blah)-[b:Foo]->(c:Goo) return p limit 1", resolve_objects= True)
    ```
This would return a Path object on `z`.

If that is the general idea then I can write the tests, if there are more details that I have missed please write here a minimal example of how to make use of the specific code you have contributed.

Do you think we can do this?

Reply:
Hey @aanastasiou,

I have made the changes according to your request. 

@mariusconjeaud 
I have rebuilt the branch from `rc/5.11` because I am not sure that the merge from rc/5.11 into my branch was made correctly at first.

We should be good now!

Reply:
Here's the cypherquery I use:

``` cypher
MATCH (p)
WITH p, rand() AS random
ORDER BY random
LIMIT 1
MATCH path = (p)-[r:links*0..2]-(connected)
RETURN p,path
```

I have multiple objects with `links` (these are relationships with a tag on them). 

The goal is to obtain the graph at a certain depth from a center point (here a random one).

Do you need anything else?

Reply:
@73VW Thank you. I am aiming to start looking on this later today or from tomorrow onwards.

All the best
AA

Reply:
@73VW Alright, this looks straightforward. 

However, I think that we have a minor clash with capabilities that were added recently.

First of all, please see [here](https://neomodel.readthedocs.io/en/latest/getting_started.html#retrieving-additional-relations) and [here](https://github.com/neo4j-contrib/neomodel/issues/709) for a brief description of that capability. The inner workings of which were largely added [via this](https://github.com/neo4j-contrib/neomodel/pull/686/commits/2ccd44a0d61773c85e204b7f94b145eb9d57064d)

My preference would be to keep only this (#716) contribution for that capability, primarily because it is based on the neo4j driver itself.

The specific example that is presented in the docs would be as simple as:

```
    q = neomodel.db.cypher_query("match p=(a:Person)-[:IS_FROM]->(b:Country) return p", resolve_objects = True)
```

And for an additional example from my testing:

```
    q = db.cypher_query("match p = (a:City)<-[r:LIVES_IN]-(b:Person)-[:IS_FROM]->(c:Country) return p", resolve_objects=True)

This makes `q` a `Path` object that is entirely browsable. And these paths can have any "shape", with any depth, etc.

Now, I know that @mariusconjeaud and @tonioo are already using the 686 pull's changes. Others might be doing so as well.

I would suggest that we have a brief discussion (in any way is convenient) to decide how to proceed with this: 

1. First of all, let's all confirm that it is worth to keep only one implementation of this feature (?) 
2. Secondly, would it be possible to keep the "syntax" of 686 and when the time comes to apply the query to "translate" it to a path? I have not given a lot of thought to this but I can also see the value of making this capability available through neomodel's querying methods, but instead of returning an ad-hoc structure, have it return a Path (?).
3. Finally consider how to proceed with including this feature in an upcoming version. 

@mariusconjeaud and @tonioo For a very minimal "test" of this capability, have a look at [this one](https://gist.github.com/aanastasiou/b89aa67869daaa0c9e135d0128ae086c)

I am aware that a few people might be on vacation around this time, let's try to have this chat as soon as everyone is available.

All the best

Reply:
Sorry, forgot to add:

In #716, the Path **needs to apply** object resolution to its nodes and relationships because at the moment it returns them as plain `Node, Relationship` objects. But this is something that is straightforward to add.

Reply:
Hey @aanastasiou, I will check that in depth during the week.

Regarding your last comment, it applies object resolution to path sub-objects, check [this](https://github.com/neo4j-contrib/neomodel/pull/716/commits/e7b4975ca7c9c0be87c70418e7e42465aa5332e6#diff-5a14848686f658a9fab17bad447b0d3631c13d078bd3bbde13804138e3150e24R312) and [this](https://github.com/neo4j-contrib/neomodel/pull/716/commits/e7b4975ca7c9c0be87c70418e7e42465aa5332e6#diff-5a14848686f658a9fab17bad447b0d3631c13d078bd3bbde13804138e3150e24R315). 
That makes the resolution recursive. 

Or am I missing something?

Reply:
@73VW , the gist I used is [here](https://gist.github.com/aanastasiou/b89aa67869daaa0c9e135d0128ae086c). This was used along with your PR.

When I examine the returned result, I am getting `neo4j.graph.Node` (i.e  `type(q[0][0][0].nodes)`). This behaviour is common at least across Neo4j 4.4.18 and 5.9.

If this is not inline with what you are getting, then please post a detailed setup (neo4j server, neomodel version / commit, etc) along with a gist that demonstrates the bare minimum of the functionality you are contributing. Feel free to use mine as a basis from above. Chances are that this minimal gist will also constitute the basis for writing the pytest tests that test that part of the code. For more information, have a loot in [this directory](https://github.com/neo4j-contrib/neomodel/tree/master/test) (See for example `test_database_management.py`).

Reply:
@73VW , any progress with this?

Reply:
Hey @aanastasiou, I'm not sure to understant what's wrong with this. Aren't we looking to obtain `neo4j.graph.Node` objects?

I might have missed something here.

Also, I didn't have time to look at the link you pasted. I have given a quick look but couldn't understand everything yet.

Reply:
Just back from holiday ; waiting for this to be ready then I'll publish version 5.1.1 :-)

Reply:
@73VW Path's initialisation does not work as intended. Please see [this](https://github.com/neo4j-contrib/neomodel/pull/738) and more importantly [this](https://github.com/neo4j-contrib/neomodel/pull/738/commits/da411b0780de2f32be744616786f77daade583fb#diff-44a3e8dfb7983646a58a76ce9f70f4251188c28c5206b4369e31fde3b4458675R1).